### PR TITLE
Add functionality allowing for file attachment of notes via the V8 API. 

### PR DESCRIPTION
**Pull request by craigpanton**
To add functionality allowing for file attachment of notes via the V8 API. 

## Description
This allows file attachment of notes via the API. It requires use of the filename (which should include file extension) and filecontents (which should be base64 encoded) parameters in the API request. Other than these two parameters it should work the same as any other insert or update API call.

## Motivation and Context
Currently this is not possible via the API.

## How To Test This
This can be tested by created API calls to create, update and delete a note. For the create and update calls, you should include the filename and filecontents parameters (ensure these are the right format as stated above). Ensure that the file is added/changed to the record in the CRM and also check the record is correct in the upload directory (the name in this directory should match the ID of the record).  For delete, ensure it deletes both the record and the file from the upload directory.

Should be tested with multiple file types. 

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

### Final checklist

- [x] My code follows the code style of this project found [here](https://docs.suitecrm.com/community/contributing-code/coding-standards/).
- [x] My change requires a change to the documentation.
- [x] I have read the [**How to Contribute**](https://docs.suitecrm.com/community/contributing-code/) guidelines.

### CODE REVIEW

File: modules/Notes/vardefs.php [hunk 1]
```diff
@@ -240,7 +240,13 @@
     'comment' => 'Record deletion indicator'
   ),
 
-
+  'filecontents' =>
```
Comment (Dillon-Brown) [inline modules/Notes/vardefs.php]: PSR-2 + Short array syntax:

```php
        'filecontents' =>
            [
                'name' => 'filecontents',
                'vname' => 'LBL_FILE_CONTENTS',
                'type' => 'varchar',
                'source' => 'non-db',
            ],
```

**Comment by samus-aran** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/samus-aran/SuiteCRM/pullRequest?prid=4737600&bid=15814260)

File: Api/V8/Service/ModuleService.php [hunk 2]
```diff
@@ -200,11 +201,25 @@ public function createRecord(CreateModuleParams $params, Request $request)
         $this->setRecordUpdateParams($bean, $attributes);
 
         foreach ($attributes as $property => $value) {
```
Comment (Dillon-Brown) [inline Api/V8/Service/ModuleService.php]: PSR-2 (Cyclomatic complexity could also potentially be reduced here, not required though as the method still meets the standard): 

```php
        foreach ($attributes as $property => $value) {
            if ($property === 'filecontents') {
                continue;
            }
            if ($property === 'filename') {
                $createFile = true;
            } else {
                $bean->$property = $value;
            }
        }

        if ($createFile) {
            $this->addFileToNote($bean, $attributes);
        }
```

File: Api/V8/Service/ModuleService.php [hunk 3]
```diff
@@ -219,6 +234,54 @@ public function createRecord(CreateModuleParams $params, Request $request)
...
+    /**
+     * @param $bean
+     * @param $attributes
+     * @return mixed
+     */
+    private function addFileToNote($bean, $attributes)
+    {
+        global $sugar_config;
+
+        if(!$bean->id) {
```
Comment (Dillon-Brown) [inline Api/V8/Service/ModuleService.php]:
```suggestion
        if (!$bean->id) {
```

File: Api/V8/Service/ModuleService.php [hunk 4]
```diff
@@ -219,6 +234,54 @@ public function createRecord(CreateModuleParams $params, Request $request)
...
+            $bean->id = create_guid();
+            $bean->new_with_id = true;
+        }
+
+        // Write temp file to upload dir
+        try {
+
+            // Checking file extension
+            $ext_pos = strrpos($attributes['filename'], ".");
```
Comment (Dillon-Brown) [inline Api/V8/Service/ModuleService.php]:
```suggestion
            $extPos = strrpos($attributes['filename'], '.');
```

File: Api/V8/Service/ModuleService.php [hunk 5]
```diff
@@ -219,6 +234,54 @@ public function createRecord(CreateModuleParams $params, Request $request)
...
+            $fileExtension = substr($attributes['filename'], $ext_pos + 1);
```
Comment (Dillon-Brown) [inline Api/V8/Service/ModuleService.php]:
```suggestion
            $fileExtension = substr($attributes['filename'], $extPos + 1);
```

File: Api/V8/Service/ModuleService.php [hunk 6]
```diff
@@ -219,6 +234,54 @@ public function createRecord(CreateModuleParams $params, Request $request)
...
+            if (in_array($fileExtension, $sugar_config['upload_badext'])) {
```
Comment (Dillon-Brown) [inline Api/V8/Service/ModuleService.php]: A third parameter should be passed in here to specify type safety.

File: Api/V8/Service/ModuleService.php [hunk 7]
```diff
@@ -219,6 +234,54 @@ public function createRecord(CreateModuleParams $params, Request $request)
...
+                echo 'This file type is not valid. Please choose a valid file type.';
+                return;
```
Comment (Dillon-Brown) [inline Api/V8/Service/ModuleService.php]: Missing a return argument.

File: Api/V8/Service/ModuleService.php [hunk 8]
```diff
@@ -219,6 +234,54 @@ public function createRecord(CreateModuleParams $params, Request $request)
...
+            }
+
+            $fileName = $bean->id;
+            $fileContents = $attributes['filecontents'];
+            $targetPath="upload/" . $fileName;
```
Comment (Dillon-Brown) [inline Api/V8/Service/ModuleService.php]:
```suggestion
            $targetPath = 'upload/' . $fileName;
```

File: Api/V8/Service/ModuleService.php [hunk 9]
```diff
@@ -219,6 +234,54 @@ public function createRecord(CreateModuleParams $params, Request $request)
...
+            $content= base64_decode($fileContents);
```
Comment (Dillon-Brown) [inline Api/V8/Service/ModuleService.php]:
```suggestion
            $content = base64_decode($fileContents);
```

File: Api/V8/Service/ModuleService.php [hunk 10]
```diff
@@ -219,6 +234,54 @@ public function createRecord(CreateModuleParams $params, Request $request)
...
+
+            $file = fopen($targetPath, 'w');
```
Comment (Dillon-Brown) [inline Api/V8/Service/ModuleService.php]: Can this be made binary safe?

File: Api/V8/Service/ModuleService.php [hunk 11]
```diff
@@ -219,6 +234,54 @@ public function createRecord(CreateModuleParams $params, Request $request)
...
+            fwrite($file, $content);
+            fclose($file);
+        }
+        catch (Exception $e) {
```
Comment (Dillon-Brown) [inline Api/V8/Service/ModuleService.php]: Undefined class

File: Api/V8/Service/ModuleService.php [hunk 12]
```diff
@@ -219,6 +234,54 @@ public function createRecord(CreateModuleParams $params, Request $request)
...
+            echo 'Unable to write to attach file: ',  $e->getMessage(), "\n";
```
Comment (Dillon-Brown) [inline Api/V8/Service/ModuleService.php]: This should be thrown or logged.

File: Api/V8/Service/ModuleService.php [hunk 13]
```diff
@@ -219,6 +234,54 @@ public function createRecord(CreateModuleParams $params, Request $request)
...
+        }
+
+        // Write temp file to upload dir
+        try {
+
+            // Checking file extension
+            $ext_pos = strrpos($attributes['filename'], ".");
+            $fileExtension = substr($attributes['filename'], $ext_pos + 1);
+            if (in_array($fileExtension, $sugar_config['upload_badext'])) {
+                echo 'This file type is not valid. Please choose a valid file type.';
```
Comment (Dillon-Brown) [inline Api/V8/Service/ModuleService.php]: This should be thrown or logged.

File: Api/V8/Service/ModuleService.php [hunk 14]
```diff
@@ -239,7 +303,23 @@ public function updateRecord(UpdateModuleParams $params, Request $request)
         $this->setRecordUpdateParams($bean, $attributes);
 
         foreach ($attributes as $property => $value) {
-            $bean->$property = $value;
+
+            if($property === 'filecontents')
```
Comment (Dillon-Brown) [inline Api/V8/Service/ModuleService.php]: PSR-2 (Perhaps this should be made into a function?):

```php
        foreach ($attributes as $property => $value) {

            if ($property === 'filecontents') {
                continue;
            }
            if ($property === 'filename') {
                $createFile = true;
            } else {
                $bean->$property = $value;
            }
        }

        if ($createFile) {
            $this->addFileToNote($bean, $attributes);
        }
```

**Comment by samus-aran** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 2
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/samus-aran/SuiteCRM/pullRequest?prid=4737600&bid=15814260)

**Comment by Mac-Rae** [comment]
Appears that when deleted the note stays within the favourites. Looks like manual deletion works as expected however

**Comment by Mac-Rae** [comment]
LGTM :+1:

**Pull request merged**